server/schema_syncer.go

```diff
@@ -39,107 +39,112 @@ func (s *SchemaSyncer) Run(ctx context.Context, wg *sync.WaitGroup) {
-							}
-						}
-					}(instance)
-				}
-			}()
-		case instance := <-instanceSyncChan:
-			go func(instance *api.Instance) {
-				defer func() {
-					if r := recover(); r != nil {
-						err, ok := r.(error)
-						if !ok {
-							err = errors.Errorf("%v", r)
-						}
-						log.Error("Schema syncer PANIC RECOVER", zap.Error(err), zap.Stack("panic-stack"))
-					}
-				}()
-
-				databaseList, err := s.server.store.FindDatabase(ctx, &api.DatabaseFind{InstanceID: &instance.ID})
-				if err != nil {
-					log.Debug("Failed to find databases for the syncing instance",
-						zap.Int("id", instance.ID),
-						zap.String("name", instance.Name),
-						zap.String("error", err.Error()))
-					return
-				}
-				for _, database := range databaseList {
-					if database.SyncStatus != api.OK {
-						continue
-					}
-
-					// If we fail to sync a particular database due to permission issue, we will continue to sync the rest of the databases.
-					if err := s.server.syncDatabaseSchema(ctx, instance, database.Name); err != nil {
-						log.Debug("Failed to sync database schema",
-							zap.Int("instanceID", instance.ID),
-							zap.String("instanceName", instance.Name),
-							zap.String("databaseName", database.Name),
-							zap.String("error", err.Error()))
-					}
-				}
-			}(instance)
+			s.syncAllInstances(ctx)
+			s.syncAllDatabases(ctx)
+		case <-allDatabasesSyncChan:
+			s.syncAllDatabases(ctx)
 		case <-ctx.Done(): // if cancel() execute
 			return
 		}
 	}
 }
+
+func (s *SchemaSyncer) syncAllInstances(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			err, ok := r.(error)
+			if !ok {
+				err = errors.Errorf("%v", r)
+			}
+			log.Error("Instance syncer PANIC RECOVER", zap.Error(err), zap.Stack("panic-stack"))
+		}
+	}()
+
+	rowStatus := api.Normal
+	instanceFind := &api.InstanceFind{
+		RowStatus: &rowStatus,
+	}
+	instanceList, err := s.server.store.FindInstance(ctx, instanceFind)
+	if err != nil {
+		log.Error("Failed to retrieve instances", zap.Error(err))
+		return
+	}
+
+	var instanceWG sync.WaitGroup
+	for _, instance := range instanceList {
+		instanceWG.Add(1)
+		go func(instance *api.Instance) {
+			defer instanceWG.Done()
+			log.Debug("Sync instance schema", zap.String("instance", instance.Name))
+			if _, err := s.server.syncInstance(ctx, instance); err != nil {
+				log.Debug("Failed to sync instance",
+					zap.Int("id", instance.ID),
+					zap.String("name", instance.Name),
+					zap.String("error", err.Error()))
+				return
+			}
+		}(instance)
+	}
+	instanceWG.Wait()
+}
+
+func (s *SchemaSyncer) syncAllDatabases(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			err, ok := r.(error)
+			if !ok {
+				err = errors.Errorf("%v", r)
+			}
+			log.Error("Database syncer PANIC RECOVER", zap.Error(err), zap.Stack("panic-stack"))
+		}
+	}()
+
+	okSyncStatus := api.OK
+	maxLastSuccessfulSyncTs := time.Now().Add(-schemaSyncInterval).Unix()
```

Comment:
This changes the logic. 

Before, we are going to sync a particular instance.

With this change, we are now syncing all databases as long as they are not synced for schemaSyncInterval.

Thus, we will

1. Observe sync delay
2. Cause excessive sync (imagine we have changed a database from a small instance, it will cause we sync other very large instances having a lot of databases)

Reply:
1. Updated syncAllDatabases to sync the particular instance only.
2. It's okay since we leverage the LastSuccessfulSyncTs. We will only databases that are not synced recently (last 30 minutes).